### PR TITLE
docs: add html report overview writeup

### PR DESCRIPTION
**Pull request by paulirish**
LS folks need this, so I figured why not document in the repo. ;)

### CODE REVIEW

File: docs/report.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+## HTML Report Generation Overview
+
+Lighthouse has an indepedent report renderer that takes the **LHR** (Lighthouse Result object) and creates a DOM tree of the report. It's all done client-side.
+
+Example standalone HTML report, delivered by the CLI: https://crazy-roll.surge.sh/dbwtester.html _(View the source! 📖)_
+
+### Report Renderer components
+
+1. [`report-generator`](https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/report/report-generator.js) is the current entry point.  It compiles together the HTML string with everything required within it
+   - It runs natively in Node.js, but can run in the browser after a compile step applied during our bundling pipeline. That compile step uses a browserify transform that replaces any `fs.readFileSync()` calls to with the file content as strings.
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply docs/report.md]: > calls to fs

no?

Comment (patrickhulce) [inline docs/report.md]: nit: no comma `in Node.js but`

File: docs/report.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+1. [`lighthouse-core/report/html/`](https://github.com/GoogleChrome/lighthouse/tree/master/lighthouse-core/report/html) has everything required to create the HTML report
+1. [`lighthouse-core/report/html/renderer`](https://github.com/GoogleChrome/lighthouse/tree/master/lighthouse-core/report/html/renderer) are all client-side JS files and transform an LHR object into a report DOM tree. They are all executed within the browser.
+1. The DOM report is kicked off clientside with these four lines:
+https://github.com/GoogleChrome/lighthouse/blob/63c999789dc08b9a3b56b22f25f478f13050da29/lighthouse-core/report/html/report-template.html#L27-L31
+
+
+### Data Hydration
+`innerHTML` is avoided, the renderer relies on basic `createElement` as well as multiple components defined in `<template>` tags that are added via [`document.importNode()`](https://developer.mozilla.org/en-US/docs/Web/API/Document/importNode) and filled in via the `querySelector`/`textContent` combo.
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply docs/report.md]: `,` -> `;` :D

File: docs/report.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+1. _Chrome DevTools Audits Panel_: It gets the `renderer` files rolled into the Chromium repo, and they execute within the DevTools context. The audits panel [receives the LHR object within a WebWorker](https://github.com/ChromeDevTools/devtools-frontend/blob/master/front_end/audits2/Audits2ProtocolService.js#L27-L35), `postMessage`'s it back to the page, and runs [the renderer within DevTools UI](https://github.com/ChromeDevTools/devtools-frontend/blob/fee00605cada877c1f8e3aae758a0f8d05b64476/front_end/audits2/Audits2Panel.js#L519-L542), making a few upgrades ([one](https://github.com/ChromeDevTools/devtools-frontend/blob/fee00605cada877c1f8e3aae758a0f8d05b64476/front_end/audits2/Audits2Panel.js#L570-L583), [two](https://github.com/ChromeDevTools/devtools-frontend/blob/fee00605cada877c1f8e3aae758a0f8d05b64476/front_end/audits2/Audits2Panel.js#L596-L637)) and [simplifications](https://github.com/ChromeDevTools/devtools-frontend/blob/fee00605cada877c1f8e3aae758a0f8d05b64476/front_end/audits2/Audits2Panel.js#L275-L304).
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply docs/report.md]: yeah page -> panel might be a little more clear

**Comment by paulirish** [comment]
lgtm plx :D

File: docs/report.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
+## HTML Report Generation Overview
+
+Lighthouse has an indepedent report renderer that takes the **LHR** (Lighthouse Result object) and creates a DOM tree of the report. It's all done client-side.
+
+Example standalone HTML report, delivered by the CLI: [`dbwtest-3.0.0-alpha.html`](https://googlechrome.github.io/lighthouse/reports/dbwtest-3.0.0-alpha.html) _(View the source! 📖)_
+
+### Report Renderer components
+
+1. [`lighthouse-core/report/report-generator.js`](https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/report/report-generator.js) is the current entry point. It compiles together the HTML string with everything required within it
```
Comment (brendankenny) [inline docs/report.md]: end with .

File: docs/report.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+   - It runs natively in Node.js but can run in the browser after a compile step is applied during our bundling pipeline. That compile step uses a browserify transform that takes any `fs.readFileSync()` calls and replaces them the stringified file content.
```
Comment (brendankenny) [inline docs/report.md]: replaces them *with* the stringified file content

File: docs/report.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+1. [`lighthouse-core/report/html/`](https://github.com/GoogleChrome/lighthouse/tree/master/lighthouse-core/report/html) has everything required to create the HTML report
```
Comment (brendankenny) [inline docs/report.md]: .

File: docs/report.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+1. [`lighthouse-core/report/html/renderer`](https://github.com/GoogleChrome/lighthouse/tree/master/lighthouse-core/report/html/renderer) are all client-side JS files and transform an LHR object into a report DOM tree. They are all executed within the browser.
```
Comment (brendankenny) [inline docs/report.md]: maybe
```
are all client-side JS files. They transform an LHR object into a report DOM tree. They are all executed within the browser.
```

File: docs/report.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+1. [`lighthouse-core/report/html/report-template.html`] is where the client-side build of the DOM report is typically kicked off ([with these four lines](https://github.com/GoogleChrome/lighthouse/blob/63c999789dc08b9a3b56b22f25f478f13050da29/lighthouse-core/report/html/report-template.html#L27-L31)) However, see _Current Uses.._ below for more possibilities.
+
+
+### Data Hydration
+`innerHTML` is deliberated not used. The renderer relies on basic `createElement` as well as multiple components defined in `<template>` tags that are added via [`document.importNode()`](https://developer.mozilla.org/en-US/docs/Web/API/Document/importNode) and filled in via the `querySelector`/`textContent` combo.
```
Comment (brendankenny) [inline docs/report.md]: deliberately

File: docs/report.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+Examples:
+
+* [`templates.html`](https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/report/html/templates.html)
+* [`performance-category-renderer.js`](https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/report/html/renderer/performance-category-renderer.js)
+
+### Current uses of report renderer
+
+The renderer was designed to be portable across various environments.
+
+1. _LH Chrome Extension_: It [creates the HTML as the runner finishes up](https://github.com/GoogleChrome/lighthouse/blob/440155cdda377c458c0efce006bc3a69ce2a351c/lighthouse-core/runner.js#L137-L138) and transforms it [into a blob url in the background page](https://github.com/GoogleChrome/lighthouse/blob/440155cdda377c458c0efce006bc3a69ce2a351c/lighthouse-extension/app/src/lighthouse-ext-background.js#L129-L143).
```
Comment (brendankenny) [inline docs/report.md]: .

File: docs/report.md [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+1. _Chrome DevTools Audits Panel_: The `renderer` are rolled into the Chromium repo, and they execute within the DevTools context. The audits panel [receives the LHR object from a WebWorker](https://github.com/ChromeDevTools/devtools-frontend/blob/master/front_end/audits2/Audits2ProtocolService.js#L27-L35), `postMessage`'s it back to the panel, and runs [the renderer within DevTools UI](https://github.com/ChromeDevTools/devtools-frontend/blob/fee00605cada877c1f8e3aae758a0f8d05b64476/front_end/audits2/Audits2Panel.js#L519-L542), making a few upgrades ([one](https://github.com/ChromeDevTools/devtools-frontend/blob/fee00605cada877c1f8e3aae758a0f8d05b64476/front_end/audits2/Audits2Panel.js#L570-L583), [two](https://github.com/ChromeDevTools/devtools-frontend/blob/fee00605cada877c1f8e3aae758a0f8d05b64476/front_end/audits2/Audits2Panel.js#L596-L637)) and [simplifications](https://github.com/ChromeDevTools/devtools-frontend/blob/fee00605cada877c1f8e3aae758a0f8d05b64476/front_end/audits2/Audits2Panel.js#L275-L304).
```
Comment (brendankenny) [inline docs/report.md]: The `renderer` *files* are?

Comment (brendankenny) [reply docs/report.md]: now it's saying that the panel gets the LHR and postmessages it to the panel :)

**Pull request merged**